src/lib.rs

```diff
@@ -11,6 +11,678 @@
+pub const SINK_NUM: &str = "4";
+pub const PRIORITY_NUM: &str = "5";
+pub const TTL_NUM: &str = "6";
+pub const PERM_LEVEL_NUM: &str = "7";
+pub const COMMSTATUS_NUM: &str = "8";
+pub const REQID_NUM: &str = "9";
+pub const TOKEN_NUM: &str = "10";
+pub const TRACEPARENT_NUM: &str = "11";
+pub const PAYLOAD_NUM: &str = "12";
+
+// URI Wildcard consts
+const WILDCARD_AUTHORITY: &str = "*";
+const WILDCARD_ENTITY_ID: u32 = 0x0000_FFFF;
+const WILDCARD_ENTITY_VERSION: u32 = 0x0000_00FF;
+const WILDCARD_RESOURCE_ID: u32 = 0x0000_FFFF;
+
+/// Configuration for the mqtt client.
+pub struct MqttConfig {
+    /// Port of the mqtt broker to connect to.
+    pub mqtt_port: String,
+    /// Hostname of the mqtt broker.
+    pub mqtt_hostname: String,
+    /// Optional SSL options for the mqtt connection.
+    pub ssl_options: Option<mqtt::SslOptions>,
+}
+
+/// UP Client for mqtt.
+pub struct UPClientMqtt {
+    mqtt_client: Arc<mqtt::AsyncClient>,
+    topic_listener_map: Arc<RwLock<HashMap<String, HashSet<ComparableListener>>>>,
+    // My authority
+    authority_name: String,
+    // Indicates where client instance is running.
+    client_type: UPClientMqttType,
+}
+
+pub enum UPClientMqttType {
+    Device,
+    Cloud,
+}
+
+#[allow(dead_code)]
+impl UPClientMqtt {
+    /// Create a new UPClientMqtt.
+    ///
+    /// # Arguments
+    /// * `config` - Configuration for the mqtt client.
+    /// * `client_id` - Client id for the mqtt client.
+    pub async fn new(
+        config: MqttConfig,
+        client_id: UUID,
+        authority_name: String,
+        client_type: UPClientMqttType,
+    ) -> Result<Self, UStatus> {
+        let topic_listener_map = Arc::new(RwLock::new(HashMap::new()));
+
+        let ptr = topic_listener_map.clone();
+
+        let mqtt_protocol = if config.ssl_options.is_some() {
+            "mqtts"
+        } else {
+            "mqtt"
+        };
+
+        let mqtt_uri = format!(
+            "{}://{}:{}",
+            mqtt_protocol, config.mqtt_hostname, config.mqtt_port
+        );
+
+        let mqtt_cli = mqtt::CreateOptionsBuilder::new()
+            .server_uri(mqtt_uri)
+            .client_id(client_id)
+            .max_buffered_messages(100)
+            .create_client()
+            .map_err(|e| {
+                UStatus::fail_with_code(
+                    UCode::INTERNAL,
+                    format!("Unable to create mqtt client: {e:?}"),
+                )
+            })?;
+
+        mqtt_cli.set_message_callback(
+            move |cli: &AsyncClient, message: std::option::Option<paho_mqtt::Message>| {
+                UPClientMqtt::on_receive(cli, message, ptr.clone())
+            },
+        );
+
+        // TODO: Integrate ssl options when connecting, may need a username, etc.
+        let conn_opts = mqtt::ConnectOptionsBuilder::with_mqtt_version(MQTT_VERSION_5)
+            .clean_start(false)
+            .properties(mqtt::properties![mqtt::PropertyCode::SessionExpiryInterval => 3600])
```

Comment:
Should SessionExpiryInterval be configurable? Not familiar enough with MQTT5 to tell

Reply:
made it configurable